File: problems/0062-unique-paths/solution.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        """
        Problem: 0062 - Unique Paths

        A robot is located at the top-left corner of an m × n grid. The robot can only move
        either right or down at each step. Return the number of unique paths to reach the
        bottom-right corner of the grid.

        Approach:
        - Use bottom-up dynamic programming.
        - Let dp[i][j] be the number of unique paths to reach cell (i, j).
        - Initialize the first row and first column with 1s (only one way to reach those).
        - For all other cells, dp[i][j] = dp[i-1][j] + dp[i][j-1].

        Time complexity: O(m * n)
        Space complexity: O(m * n), can be reduced to O(n) with 1D DP.
        """
        dp = [[1] * n for _ in range(m)]  # Initialize dp table with 1s.

        # Fill the table row by row.
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1]  # From top + from left.

        return dp[m - 1][n - 1]  # Number of unique paths to bottom-right corner.

File: problems/0062-unique-paths/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    /*
    Problem: 0062 - Unique Paths

    A robot is located at the top-left corner of an m × n grid. The robot can only move
    either right or down at each step. Return the number of unique paths to reach the
    bottom-right corner of the grid.

    Approach:
    - Use bottom-up dynamic programming.
    - Let dp[i][j] store the number of unique paths to reach cell (i, j).
    - Initialize all cells in the first row and first column to 1 (only one way to reach them).
    - For all other cells, compute: dp[i][j] = dp[i-1][j] + dp[i][j-1].

    Time Complexity: O(m * n)
    Space Complexity: O(m * n), can be optimized to O(n) using 1D DP.
    */
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m, vector<int>(n, 1));  // Initialize dp table with 1s.

        // Fill the table row by row.
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];  // From top + from left.
            }
        }

        return dp[m - 1][n - 1];  // Return number of paths to bottom-right.
    }
};

